### PR TITLE
ZBUG-2738: Create a hash of the key in Nginx instead of raw value

### PR DESCRIPTION
**Pull request by inpati**
**Problem statement**
Create a hash of the key in Nginx instead of raw value

**Solution**
Hashed raw key text using hashing algorithm before putting key-value in cache

### CODE REVIEW

**Review by suchitgupta11 — APPROVED** [review]

File: src/core/ngx_memcache.c [hunk 1]
```diff
@@ -1984,6 +2005,27 @@ static ngx_uint_t ngx_memcache_perl_hash (u_char *key, size_t len)
...
+    u_char digest[SHA256_KEY_LENGTH];
+    SHA256(key, len, digest);
+
+    /*
+     * this will give you back an integer representation of the hash,
+     * you can then turn this into a hex representation,
+     * if you want to pass it around as a string.
+     */
+
+    static u_char hashed_key[SHA256_KEY_LENGTH+1];
```
Comment (simon-scannell-sonarsource) [inline src/core/ngx_memcache.c]: I would make this variable dynamically allocated on the heap just to be sure. I'm not sure if this is an actual issue but using a static variable (essentially a global) might lead to race conditions.

**Review by k-kato — APPROVED** [review]

File: src/core/ngx_memcache.c (hunk 1 above)
Comment (inpati) [reply src/core/ngx_memcache.c]: Done

File: src/core/ngx_memcache.c [hunk 2]
```diff
@@ -1984,6 +2002,17 @@ static ngx_uint_t ngx_memcache_perl_hash (u_char *key, size_t len)
     return h;
 }
 
+static ngx_str_t ngx_sha256_hash (ngx_pool_t* p, u_char* key, size_t len)
+{
+    ngx_str_t hashed_key;
+    u_char * digest = ngx_pnalloc(p, SHA256_KEY_LENGTH);
+    SHA256(key, len, digest);
+    hashed_key.data = ngx_memcache_hexstr(digest, SHA256_KEY_LENGTH);
```
Comment (desouzas) [inline src/core/ngx_memcache.c:2010]: If the allocation fails and returns null this or the next line will :boom:

So the call site's null check won't be reached. I see similar on the other md5 method in this file so probably not a big deal.

**Review by desouzas — APPROVED** [review]

File: src/core/ngx_memcache.c (hunk 2 above)
Comment (inpati) [reply src/core/ngx_memcache.c:2010]: Ok. understood your point ignoring boom for now as done in md5 !

**Pull request merged**